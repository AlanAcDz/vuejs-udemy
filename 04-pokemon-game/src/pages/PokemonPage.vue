<template>
    <template v-if="pokemon">
        <h1>¿Quién es este pokemon?</h1>
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
        <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer($event)" />
        <template v-if="showAnswer">
            <h2 class="fade-in">{{ message }}</h2>
            <button @click="newGame()">Nuevo Juego</button>
        </template>
    </template>
    <h1 v-else>Espere por favor...</h1>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'
import { getPokemonOptions } from '@/helpers/getPokemonOptions'

export default {
    components: { PokemonPicture, PokemonOptions },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: '',
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()
            const rndInt = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr.at(rndInt)
        },
        checkAnswer(selectedId) {
            if (this.showAnswer) return
            this.showAnswer = true
            this.showPokemon = true
            this.message =
                selectedId === this.pokemon.id ? `Correcto, ${this.pokemon.name}` : `Oops, era ${this.pokemon.name}`
        },
        newGame() {
            this.pokemon = null
            this.showPokemon = false
            this.showAnswer = false
            this.message = ''
            this.mixPokemonArray()
        },
    },
    mounted() {
        this.mixPokemonArray()
    },
}
</script>
