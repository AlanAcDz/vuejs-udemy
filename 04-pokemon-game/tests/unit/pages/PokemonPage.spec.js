import { shallowMount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { mockPokemons } from '../mocks/pokemons.mock'

const mockData = {
    pokemonArr: mockPokemons,
    pokemon: mockPokemons[0],
    showPokemon: false,
    showAnswer: false,
    message: '',
}

describe('PokemonPage component', () => {
    let wrapper
    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })
    test('should match snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
    test('should call mixPokemonArray on mount', () => {
        const mixPokemoArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        shallowMount(PokemonPage)
        expect(mixPokemoArraySpy).toHaveBeenCalled()
    })
    test('should match snapshot when pokemons load', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return mockData
            },
        })
        expect(wrapper.html()).toMatchSnapshot()
    })
    test('should show PokemonPicture & PokemonOptions components', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return mockData
            },
        })
        const pokemonPicture = wrapper.find('pokemon-picture-stub')
        const pokemonOptions = wrapper.find('pokemon-options-stub')
        expect(pokemonPicture.exists()).toBeTruthy()
        expect(pokemonOptions.exists()).toBeTruthy()
        expect(pokemonPicture.attributes('pokemonid')).toBe('1')
        expect(pokemonOptions.attributes('pokemons')).toBeTruthy()
    })
    test('checkAnswer tests', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return mockData
            },
        })
        await wrapper.vm.checkAnswer(1)
        const h2Tag = wrapper.find('h2')
        expect(h2Tag.exists()).toBeTruthy()
        expect(h2Tag.text()).toBe(`Correcto, ${mockPokemons[0].name}`)
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        // await wrapper.vm.newGame()
        // await wrapper.vm.checkAnswer(2)
        // expect(wrapper.vm.message).toBe(`Oops, era ${mockPokemons[0].name}`)
    })
})
