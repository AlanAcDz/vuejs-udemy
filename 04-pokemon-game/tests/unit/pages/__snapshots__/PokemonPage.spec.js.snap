// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage component should match snapshot when pokemons load 1`] = `
<h1>¿Quién es este pokemon?</h1>
<pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
<pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
<!--v-if-->
`;
